In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [3]:
df = pd.read_excel(r'nuclear-energy-generation.xlsx', sheet_name = 0)
df_backup = df.copy()

values = ['World']
df = df[df.Entity.isin(values) == False]

In [4]:
df.info()

df

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2052 entries, 0 to 2051
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Entity                2052 non-null   object 
 1   Code                  2052 non-null   object 
 2   Year                  2052 non-null   int64  
 3   Nuclear_Electricity   2052 non-null   float64
 4   Annual_CO2_emissions  2052 non-null   float64
dtypes: float64(2), int64(1), object(2)
memory usage: 96.2+ KB


Entity Code  Year  Nuclear_Electricity  Annual_CO2_emissions
0         Argentina  ARG  1965                 0.00              2.666444
1         Argentina  ARG  1966                 0.00              2.815352
2         Argentina  ARG  1967                 0.00              2.877178
3         Argentina  ARG  1968                 0.00              2.985735
4         Argentina  ARG  1969                 0.00              3.291103
...             ...  ...   ...                  ...                   ...
2047  United States  USA  2017               804.95             15.800776
2048  United States  USA  2018               807.08             16.187923
2049  United States  USA  2019               809.41             15.730885
2050  United States  USA  2020               789.88             14.037217
2051  United States  USA  2021               778.19             14.858668

[2052 rows x 5 columns]

In [5]:
df.Entity.value_counts()


Argentina               57
Armenia                 57
Netherlands             57
Pakistan                57
Romania                 57
Russia                  57
Slovakia                57
Slovenia                57
South Africa            57
South Korea             57
Spain                   57
Sweden                  57
Switzerland             57
Taiwan                  57
Ukraine                 57
United Arab Emirates    57
United Kingdom          57
Mexico                  57
Lithuania               57
Kazakhstan              57
Czechia                 57
Belarus                 57
Belgium                 57
Brazil                  57
Bulgaria                57
Canada                  57
China                   57
Finland                 57
Japan                   57
France                  57
Germany                 57
Hungary                 57
India                   57
Iran                    57
Italy                   57
United States           57
Name: Entity, dtype: int64

In [ ]:
AfricaList = ['South Africa']
Africa = df.loc[df['Entity'].isin(AfricaList)]

Africa

Entity Code  Year  Nuclear_Electricity  Annual_CO2_emissions
1482  South Africa  ZAF  1965                0.000              6.678414
1483  South Africa  ZAF  1966                0.000              6.479857
1484  South Africa  ZAF  1967                0.000              6.553437
1485  South Africa  ZAF  1968                0.000              6.553454
1486  South Africa  ZAF  1969                0.000              6.595164
1487  South Africa  ZAF  1970                0.000              6.688847
1488  South Africa  ZAF  1971                0.000              7.311893
1489  South Africa  ZAF  1972                0.000              7.239002
1490  South Africa  ZAF  1973                0.000              7.109694
1491  South Africa  ZAF  1974                0.000              7.040083
1492  South Africa  ZAF  1975                0.000              7.176719
1493  South Africa  ZAF  1976                0.000              7.284851
1494  South Africa  ZAF  1977                0.000              7.343225
1495  South Africa  ZAF  1978                0.000              7.224443
1496  South Africa  ZAF  1979                0.000              7.619857
1497  South Africa  ZAF  1980                0.000              7.745089
1498  South Africa  ZAF  1981                0.000              8.503257
1499  South Africa  ZAF  1982                0.000              9.039699
1500  South Africa  ZAF  1983                0.000              9.160536
1501  South Africa  ZAF  1984                3.925              9.631073
1502  South Africa  ZAF  1985                5.315              9.595049
1503  South Africa  ZAF  1986                8.803              9.475847
1504  South Africa  ZAF  1987                6.167              9.099414
1505  South Africa  ZAF  1988               10.493              9.163398
1506  South Africa  ZAF  1989               11.099              8.811032
1507  South Africa  ZAF  1990                8.449              7.847275
1508  South Africa  ZAF  1991                9.144              7.969176
1509  South Africa  ZAF  1992                9.288              7.215516
1510  South Africa  ZAF  1993                7.255              7.533027
1511  South Africa  ZAF  1994                9.697              7.825034
1512  South Africa  ZAF  1995               11.301              8.220288
1513  South Africa  ZAF  1996               11.775              8.141431
1514  South Africa  ZAF  1997               12.647              8.502075
1515  South Africa  ZAF  1998               13.601              8.222434
1516  South Africa  ZAF  1999               12.837              8.081300
1517  South Africa  ZAF  2000               13.010              8.075817
1518  South Africa  ZAF  2001               10.720              7.863059
1519  South Africa  ZAF  2002               11.990              7.475733
1520  South Africa  ZAF  2003               12.660              8.400885
1521  South Africa  ZAF  2004               14.280              9.247540
1522  South Africa  ZAF  2005               12.240              8.485409
1523  South Africa  ZAF  2006               10.070              9.022819
1524  South Africa  ZAF  2007               12.600              9.298462
1525  South Africa  ZAF  2008               12.750              9.789216
1526  South Africa  ZAF  2009               11.570              9.388231
1527  South Africa  ZAF  2010               12.900              8.937098
1528  South Africa  ZAF  2011               12.940              8.881489
1529  South Africa  ZAF  2012               12.400              8.635881
1530  South Africa  ZAF  2013               13.610              8.513041
1531  South Africa  ZAF  2014               14.760              8.621777
1532  South Africa  ZAF  2015               10.970              7.985844
1533  South Africa  ZAF  2016               15.210              8.099157
1534  South Africa  ZAF  2017               15.090              7.757252
1535  South Africa  ZAF  2018               10.560              7.5905

In [ ]:
EuropeList  =['Netherlands','Romania','Russia','Slovakia','Slovenia','Spain','Sweden','Switzerland','Ukraine','United Kingdom','Lithuania','Czechia','Belarus','Belgium','Bulgaria','Finland','France','Germany','Hungary ']
Europe = df.loc[df['Entity'].isin(EuropeList)]

Europe

Entity Code  Year  Nuclear_Electricity  Annual_CO2_emissions
114          Belarus  BLR  1965             1.860567              5.959548
115          Belarus  BLR  1966             2.063860              6.247323
116          Belarus  BLR  1967             2.253894              6.489499
117          Belarus  BLR  1968             3.133355              6.629290
118          Belarus  BLR  1969             3.632747              6.860397
...              ...  ...   ...                  ...                   ...
1990  United Kingdom  GBR  2017            70.340000              5.863442
1991  United Kingdom  GBR  2018            65.060000              5.715981
1992  United Kingdom  GBR  2019            56.180000              5.462123
1993  United Kingdom  GBR  2020            50.850000              4.865282
1994  United Kingdom  GBR  2021            46.880000              5.154100

[1026 rows x 5 columns]

In [ ]:
AmericaList  =['Argentina','Mexico','Brazil','Canada','United States']
America = df.loc[df['Entity'].isin(AmericaList)]

America

Entity Code  Year  Nuclear_Electricity  Annual_CO2_emissions
0         Argentina  ARG  1965                 0.00              2.666444
1         Argentina  ARG  1966                 0.00              2.815352
2         Argentina  ARG  1967                 0.00              2.877178
3         Argentina  ARG  1968                 0.00              2.985735
4         Argentina  ARG  1969                 0.00              3.291103
...             ...  ...   ...                  ...                   ...
2047  United States  USA  2017               804.95             15.800776
2048  United States  USA  2018               807.08             16.187923
2049  United States  USA  2019               809.41             15.730885
2050  United States  USA  2020               789.88             14.037217
2051  United States  USA  2021               778.19             14.858668

[285 rows x 5 columns]

In [ ]:
AsiaList  =  ['South Korea','Taiwan','China','Japan','India','Iran','Armenia','Pakistan','United Arab Emirates','Kazakhstan']
Asia = df.loc[df['Entity'].isin(AsiaList)]

Asia

Entity Code  Year  Nuclear_Electricity  \
57                 Armenia  ARM  1965                 0.00   
58                 Armenia  ARM  1966                 0.00   
59                 Armenia  ARM  1967                 0.00   
60                 Armenia  ARM  1968                 0.00   
61                 Armenia  ARM  1969                 0.00   
...                    ...  ...   ...                  ...   
1933  United Arab Emirates  ARE  2017                 0.00   
1934  United Arab Emirates  ARE  2018                 0.00   
1935  United Arab Emirates  ARE  2019                 0.00   
1936  United Arab Emirates  ARE  2020                 1.56   
1937  United Arab Emirates  ARE  2021                 1.79   

      Annual_CO2_emissions  
57                1.860922  
58                1.913391  
59                1.951582  
60                1.958436  
61                1.991900  
...                    ...  
1933             23.615900  
1934             22.999390  
1935             22.633892  
1936             21.436235  
1937             21.792141  

[570 rows x 5 columns]

In [ ]:
globalPowers =  ['United States','China','India']
gp = df.loc[df['Entity'].isin(globalPowers)]
gp

Entity Code  Year  Nuclear_Electricity  Annual_CO2_emissions
399           China  CHN  1965                 0.00              0.691150
400           China  CHN  1966                 0.00              0.739566
401           China  CHN  1967                 0.00              0.604760
402           China  CHN  1968                 0.00              0.634962
403           China  CHN  1969                 0.00              0.758248
...             ...  ...   ...                  ...                   ...
2047  United States  USA  2017               804.95             15.800776
2048  United States  USA  2018               807.08             16.187923
2049  United States  USA  2019               809.41             15.730885
2050  United States  USA  2020               789.88             14.037217
2051  United States  USA  2021               778.19             14.858668

[171 rows x 5 columns]

In [ ]:
import pandas as pd
from statsmodels.stats.anova import AnovaRM

# ANOVARM Europa

Hipótesis nula:

 - No hay diferencias significativas en el uso de energía nuclear entre países de Europa desde 1965 hasta 2021.

Hipótesis alternativas:

 - Hay diferencias significativas en el uso de energía nuclear entre países de Europa desde 1965 hasta 2021.

 Regla de aceptación si  el valor  de p.value es mayor al 0.05  la hipótesis  nula sera aceptada.

In [ ]:
aov = AnovaRM(Europe, 'Nuclear_Electricity', 'Entity', within=['Year'])
res = aov.fit()
print(res.summary())

               Anova
     F Value  Num DF  Den DF  Pr > F
------------------------------------
Year  5.1356 56.0000 952.0000 0.0000



Decisión: la hipótesis nula  se rechaza ya que  el  valor  de  p.value es menor al 0.05.

### Conclusión:Hay diferencias significativas en el uso de energía nuclear entre países de Europa desde 1965 hasta 2021.

# ANOVARM América  

Hipótesis nula:

 - No hay diferencias significativas en el uso de energía nuclear entre países de América desde 1965 hasta 2021.

Hipótesis alternativas:

 - Hay diferencias significativas en el uso de energía nuclear entre países de América desde 1965 hasta 2021.

Regla de aceptación si  el valor  de p.value es mayor al 0.05  la hipótesis  nula sera aceptada.

In [ ]:
aov = AnovaRM(America, 'Nuclear_Electricity', 'Entity', within=['Year'])
res = aov.fit()
print(res.summary())

               Anova
     F Value  Num DF  Den DF  Pr > F
------------------------------------
Year  1.4189 56.0000 224.0000 0.0399



Decisión: la hipótesis nula  se rechaza ya que  el  valor  de  p.value es menor al 0.05.

### Conclusión:Hay diferencias significativas en el uso de energía nuclear entre países de America desde 1965 hasta 2021.

# ANOVARM Asia 

Hipótesis nula:

 - No hay diferencias significativas en el uso de energía nuclear entre países de Asia desde 1965 hasta 2021.

Hipótesis alternativas:

 - Hay diferencias significativas en el uso de energía nuclear entre países de Asia desde 1965 hasta 2021.

Regla de aceptación si  el valor  de p.value es mayor al 0.05  la hipótesis  nula sera aceptada.

In [ ]:
aov = AnovaRM(Asia, 'Nuclear_Electricity', 'Entity', within=['Year'])
res = aov.fit()
print(res.summary())

               Anova
     F Value  Num DF  Den DF  Pr > F
------------------------------------
Year  1.4561 56.0000 504.0000 0.0210



Decisión: la hipótesis nula  se rechaza ya que  el  valor  de  p.value es menor al 0.05.

### Conclusión:Hay diferencias significativas en el uso de energía nuclear entre países de Asia desde 1965 hasta 2021.

# Linear Mixed Effects Models

### Preguntas de investigación que planeamos resolver con este método

- ¿Existe una tendencia significativa en la generación de energía ni a lo largo del tiempo por continente?
- ¿Existe una tendencia significativa en las emisiones anuales de CO2 a lo largo del tiempo por continente?


In [ ]:
dfContinents = pd.read_csv('./continents-according-to-our-world-in-data.csv',encoding='utf-8') 
dfContinents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 285 entries, 0 to 284
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Entity     285 non-null    object
 1   Code       285 non-null    object
 2   Year       285 non-null    int64 
 3   Continent  285 non-null    object
dtypes: int64(1), object(3)
memory usage: 9.0+ KB


In [ ]:
dict_countries = dfContinents[['Entity','Continent']].set_index('Entity')
dict_countries

Continent
Entity                         
Abkhazia                   Asia
Afghanistan                Asia
Akrotiri and Dhekelia      Asia
Albania                  Europe
Algeria                  Africa
...                         ...
Yugoslavia               Europe
Zambia                   Africa
Zanzibar                 Africa
Zimbabwe                 Africa
Åland Islands            Europe

[285 rows x 1 columns]

In [ ]:
dict_countries = dict_countries['Continent'].to_dict()
dict_countries

{'Abkhazia': 'Asia',
 'Afghanistan': 'Asia',
 'Akrotiri and Dhekelia': 'Asia',
 'Albania': 'Europe',
 'Algeria': 'Africa',
 'American Samoa': 'Oceania',
 'Andorra': 'Europe',
 'Angola': 'Africa',
 'Anguilla': 'North America',
 'Antarctica': 'Antarctica',
 'Antigua and Barbuda': 'North America',
 'Argentina': 'South America',
 'Armenia': 'Asia',
 'Aruba': 'North America',
 'Australia': 'Oceania',
 'Austria': 'Europe',
 'Austria-Hungary': 'Europe',
 'Azerbaijan': 'Asia',
 'Baden': 'Europe',
 'Bahamas': 'North America',
 'Bahrain': 'Asia',
 'Bangladesh': 'Asia',
 'Barbados': 'North America',
 'Bavaria': 'Europe',
 'Belarus': 'Europe',
 'Belgium': 'Europe',
 'Belize': 'North America',
 'Benin': 'Africa',
 'Bermuda': 'North America',
 'Bhutan': 'Asia',
 'Bolivia': 'South America',
 'Bonaire Sint Eustatius and Saba': 'North America',
 'Bosnia and Herzegovina': 'Europe',
 'Botswana': 'Africa',
 'Bouvet Island': 'Antarctica',
 'Brazil': 'South America',
 'British Indian Ocean Territory': 'Asia

In [ ]:
df['Continent'] = df['Entity'].map(dict_countries,na_action='ignore')
df[['Continent','Entity']]

Continent         Entity
0     South America      Argentina
1     South America      Argentina
2     South America      Argentina
3     South America      Argentina
4     South America      Argentina
...             ...            ...
2047  North America  United States
2048  North America  United States
2049  North America  United States
2050  North America  United States
2051  North America  United States

[2052 rows x 2 columns]

In [ ]:
import statsmodels.formula.api as smf
md = smf.mixedlm("Nuclear_Electricity ~ Year", df, groups=df["Continent"]) 
mdf = md.fit(method=["lbfgs"]) 
print(mdf.summary())

              Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: Nuclear_Electricity
No. Observations: 2052    Method:             REML               
No. Groups:       5       Scale:              10706.2656         
Min. group size:  57      Log-Likelihood:     inf                
Max. group size:  1140    Converged:          Yes                
Mean group size:  410.4                                          
-------------------------------------------------------------------
               Coef.    Std.Err.     z      P>|z|   [0.025   0.975]
-------------------------------------------------------------------
Intercept      -0.000                                              
Year            1.481      0.139   10.664   0.000    1.208    1.753
Group Var       0.000                                              



c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:1634: UserWarning: Random effects covariance is singular
  warnings.warn(msg)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2054: UserWarning: The random effects covariance matrix is singular.
  warnings.warn(_warn_cov_sing)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2237: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2245: UserWarning: The random effects covariance matrix is singular.
  warnings.warn(_warn_cov_sing)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2261: Converg

# Conclusiones a partir de los datos otorgados por el modelo
- El modelo indica que la producción de electricidad nuclear no presenta una intercepción significativa en el año cero, lo que podría indicar que la producción de electricidad nuclear en ese momento era cercana a cero.

- El coeficiente de año indica que, en promedio, la producción de electricidad nuclear aumentó en 1.481 unidades por año (esto es significativo, con un valor p < 0.05).

- El modelo también indica que la varianza de grupo es cero, lo que significa que no hay variación en la producción de electricidad nuclear entre los diferentes continentes.

In [ ]:

md = smf.mixedlm("Annual_CO2_emissions ~ Year", df, groups=df["Continent"]) 
mdf = md.fit(method=["lbfgs"]) 
print(mdf.summary())

              Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: Annual_CO2_emissions
No. Observations: 2052    Method:             REML                
No. Groups:       5       Scale:              39.3611             
Min. group size:  57      Log-Likelihood:     inf                 
Max. group size:  1140    Converged:          Yes                 
Mean group size:  410.4                                           
--------------------------------------------------------------------
                Coef.    Std.Err.     z      P>|z|   [0.025   0.975]
--------------------------------------------------------------------
Intercept        0.000                                              
Year            -0.018      0.008   -2.089   0.037   -0.034   -0.001
Group Var        0.000                                              



c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:1634: UserWarning: Random effects covariance is singular
  warnings.warn(msg)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2054: UserWarning: The random effects covariance matrix is singular.
  warnings.warn(_warn_cov_sing)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2237: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2245: UserWarning: The random effects covariance matrix is singular.
  warnings.warn(_warn_cov_sing)
c:\Users\david\AppData\Local\Programs\Python\Python39\lib\site-packages\statsmodels\regression\mixed_linear_model.py:2261: Converg

# Conclusiones a partir de los datos otorgados por el modelo

- El coeficiente del año (Year) es -0.018, lo que significa que, en promedio, las emisiones de dióxido de carbono disminuyen en 0.018 toneladas por año. 

- El modelo también indica que la varianza de grupo es cero, o que indica que no hay diferencias significativas en las emisiones de dióxido de carbono entre los cinco continentes.